Fixes generation of UniformMatrix4fv to match expected function signiture

kdomasze opened the pull request:
When passing a 4x4 matrix to a shader through a uniform, "glUniformMatrix4fv" expects a "float[]" as input. However, the current implementation of "GL.UniformMatrix4fv" overrides this and expects a reference to a "float". This makes the function unusable as far as I could tell (and is different from the implementations of  other "uniform" functions in this library).

In the "Generator" project, there is an override for "glUniformMatrix4fv" to change the value to "ref float". Removing these two lines allows the generator to generate the expected function parameters.

kdomasze committed:
Fixes generation of UniformMatrix4fv to match expected function signiture

When passing a 4x4 matrix to a shader through a uniform, "glUniformMatrix4fv" expects a "float[]" as input. However, the current implementation of "GL.UniformMatrix4fv" overrides this and expects a reference to a "float". This makes the function unusable as far as I could tell (and is completely different from the implementation by opengl and the other functions in this library).

In the "Generator" project, there is an override for "glUniformMatrix4fv" to change the value to this "ref float". Removing these two lines allows the generator to generate the correct function parameters.

smack0007 commented:
First off thank you for taking an interest in my project and creating a pull request. It's nice to know people are actually interested in my project.

The fact of the matter is you actually need both function signatures. Why do you need the ref overload? That's for passing struct types:

```c#
struct Matrix4
{
    public float M11;
    public float M12;
    // ...
}
```

This allows you to do `UniformMatrix4fv(ref matrix.M11);`.

This mistake that I've made here is that I implemented it the way I did solely because I needed it that way in another unrelated project. The correct way to implement this would be to clone the function in the generator and in the cloned function override the parameter to be a ref parameter. If you have the desire or the time to do it this way, I will accept the pull request. If not, I will look into when I have time but that could be a few days.

Again, thank you for taking an interest in my project.

kdomasze commented:
Thank you for the feedback and the explanation. Your explanation makes a lot of sense and I understand why it was implemented that way now.

I am interested in implementing the functionality, and just had a question about implementation to ensure I do it in a fashion you would be happy with.

From reading through the generator code, it looks like it outputs delegates for each function linking to GL. From what I understand, you can't overload delegates in C#, so, it would need a uniquely named delegate to allow for both parameters. Would you prefer a function like `UniformMatrix4fv` to have an appended descriptor (i.e. `UniformMatrix4fv(float value)` and `UniformMatrix4fvByRef(ref float value)`) or in some other fashion? 

I do have it working in this fashion right now locally and can update my pull request with it if you would be happy with that. Otherwise, I'm happy to adjust as you require.

smack0007 commented:
The `UniformMatrix4fvByRef(ref float value)` variation sounds good. I will gladly accept the pull request.

kdomasze committed:
Enables generation of copies of functions with appended descriptors

Modifies the generator to enable generation of new functions with unique names to link to already used OpenGL functions. This allows creation of functions with parameters that are not identical to what is expected by OpenGL, but still valid.

Currently the solution is hard coded to "ByRef" but could be easily adapted to support a list of possible appended names for other functions.

kdomasze commented:
I have pushed the modifications to my branch. If there are any changes you would like me to make to my implementation, let me know.
